### PR TITLE
=typ #23697 start typed receptionist eagerly

### LINKED ISSUE #23697: Receptionist extension should be loaded eagerly
So that the distributed registry is populated for early `Find` requests.

### PR DESCRIPTION
**Pull request by jrudolph**
Fixes #23697

### CODE REVIEW

**Review by ktoso — APPROVED** [review]
LGTM

**Review by johanandren — COMMENTED** [review]
I think there may be something dangerous here with the remote receptionist being eagerly initialized, touching ddata, which touches cluster, which touches remoting. I think that should fail the tests though.

**Review by chbatey — APPROVED** [review]
LGTM if the tests pass :)

**Comment by patriknw** [comment]
`Cluster` is started by `ClusterActorRefProvider` anyway, so I think this should be ok

**Comment by jrudolph** [comment]
> I think there may be something dangerous here with the remote receptionist being eagerly initialized, touching ddata, which touches cluster, which touches remoting. I think that should fail the tests though.

Good hunch, @johanandren, it seems this exactly is happening: failing because `RemoteActorRefProvider.init` has not yet been called when the receptionist is started up. Does that mean we cannot use the extensions mechanism or is there a trick to make sure everthing is initialized?

**Comment by jrudolph** [comment]
There seems to be a cycle during initialization of the `ClusterActorRefProvider`:

```
	at akka.actor.typed.receptionist.Receptionist.<init>(Receptionist.scala:24)
	at akka.actor.typed.receptionist.Receptionist$.createExtension(Receptionist.scala:82)
	at akka.actor.typed.receptionist.Receptionist$.createExtension(Receptionist.scala:81)
	at akka.actor.typed.internal.ExtensionsImpl$class.createExtensionInstance(ExtensionsImpl.scala:84)
	at akka.actor.typed.internal.ExtensionsImpl$class.registerExtension(ExtensionsImpl.scala:75)
	at akka.actor.typed.internal.adapter.ActorSystemAdapter.registerExtension(ActorSystemAdapter.scala:29)
	at akka.actor.typed.internal.ExtensionsImpl$$anonfun$loadExtensions$1$1.apply(ExtensionsImpl.scala:42)
	at akka.actor.typed.internal.ExtensionsImpl$$anonfun$loadExtensions$1$1.apply(ExtensionsImpl.scala:36)
	at scala.collection.Iterator$class.foreach(Iterator.scala:891)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1334)
	at scala.collection.IterableLike$class.foreach(IterableLike.scala:72)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at akka.actor.typed.internal.ExtensionsImpl$class.loadExtensions$1(ExtensionsImpl.scala:36)
	at akka.actor.typed.internal.ExtensionsImpl$class.loadExtensions(ExtensionsImpl.scala:62)
	at akka.actor.typed.internal.adapter.ActorSystemAdapter.loadExtensions(ActorSystemAdapter.scala:29)
	at akka.actor.typed.internal.adapter.ActorSystemAdapter.<init>(ActorSystemAdapter.scala:32)
	at akka.actor.typed.internal.adapter.ActorSystemAdapter$AdapterExtension.<init>(ActorSystemAdapter.scala:94)
	at akka.actor.typed.internal.adapter.ActorSystemAdapter$AdapterExtension$.createExtension(ActorSystemAdapter.scala:101)
	at akka.actor.typed.internal.adapter.ActorSystemAdapter$AdapterExtension$.createExtension(ActorSystemAdapter.scala:97)
	at akka.actor.ActorSystemImpl.registerExtension(ActorSystem.scala:893)
	at akka.actor.ExtensionId$class.apply(Extension.scala:78)
	at akka.actor.typed.internal.adapter.ActorSystemAdapter$AdapterExtension$.apply(ActorSystemAdapter.scala:97)
	at akka.actor.typed.internal.adapter.ActorSystemAdapter$.apply(ActorSystemAdapter.scala:90)
	at akka.actor.typed.scaladsl.adapter.AdapterExtension.<init>(AdapterExtension.scala:16)
	at akka.actor.typed.scaladsl.adapter.AdapterExtension$.createExtension(AdapterExtension.scala:22)
	at akka.actor.typed.scaladsl.adapter.AdapterExtension$.createExtension(AdapterExtension.scala:21)
	at akka.actor.ActorSystemImpl.registerExtension(ActorSystem.scala:893)
	at akka.actor.ExtensionId$class.apply(Extension.scala:78)
	at akka.actor.typed.scaladsl.adapter.AdapterExtension$.apply(AdapterExtension.scala:21)
	at akka.actor.typed.scaladsl.adapter.package$UntypedActorSystemOps$.toTyped$extension(package.scala:57)
	at akka.actor.typed.internal.MiscMessageSerializer.<init>(MiscMessageSerializer.scala:17)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at akka.actor.ReflectiveDynamicAccess$$anonfun$createInstanceFor$2.apply(ReflectiveDynamicAccess.scala:32)
	at scala.util.Try$.apply(Try.scala:192)
	at akka.actor.ReflectiveDynamicAccess.createInstanceFor(ReflectiveDynamicAccess.scala:27)
	at akka.actor.ReflectiveDynamicAccess$$anonfun$createInstanceFor$3.apply(ReflectiveDynamicAccess.scala:38)
	at akka.actor.ReflectiveDynamicAccess$$anonfun$createInstanceFor$3.apply(ReflectiveDynamicAccess.scala:38)
	at scala.util.Success.flatMap(Try.scala:231)
	at akka.actor.ReflectiveDynamicAccess.createInstanceFor(ReflectiveDynamicAccess.scala:38)
	at akka.serialization.Serialization.serializerOf(Serialization.scala:301)
	at akka.serialization.Serialization$$anonfun$6.apply(Serialization.scala:327)
	at akka.serialization.Serialization$$anonfun$6.apply(Serialization.scala:327)
	at scala.collection.TraversableLike$WithFilter$$anonfun$map$2.apply(TraversableLike.scala:683)
	at scala.collection.immutable.HashMap$HashMap1.foreach(HashMap.scala:221)
	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:428)
	at scala.collection.immutable.HashMap$HashTrieMap.foreach(HashMap.scala:428)
	at scala.collection.TraversableLike$WithFilter.map(TraversableLike.scala:682)
	at akka.serialization.Serialization.<init>(Serialization.scala:327)
	at akka.serialization.SerializationExtension$.createExtension(SerializationExtension.scala:15)
	at akka.serialization.SerializationExtension$.createExtension(SerializationExtension.scala:12)
	at akka.actor.ActorSystemImpl.registerExtension(ActorSystem.scala:893)
	at akka.actor.ExtensionId$class.apply(Extension.scala:78)
	at akka.serialization.SerializationExtension$.apply(SerializationExtension.scala:12)
	at akka.remote.RemoteActorRefProvider.init(RemoteActorRefProvider.scala:203)
	at akka.cluster.ClusterActorRefProvider.init(ClusterActorRefProvider.scala:31)
	at akka.actor.ActorSystemImpl.liftedTree2$1(ActorSystem.scala:809)
	at akka.actor.ActorSystemImpl._start$lzycompute(ActorSystem.scala:806)
	at akka.actor.ActorSystemImpl._start(ActorSystem.scala:806)
	at akka.actor.ActorSystemImpl.start(ActorSystem.scala:822)
	at akka.actor.ActorSystem$.apply(ActorSystem.scala:245)
	at akka.actor.ActorSystem$.apply(ActorSystem.scala:288)
	at akka.actor.ActorSystem$.apply(ActorSystem.scala:263)
	at akka.testkit.AkkaSpec.<init>(AkkaSpec.scala:65)
	at akka.cluster.typed.RemoteMessageSpec.<init>(RemoteMessageSpec.scala:62)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:435)
	at sbt.ForkMain$Run$2.call(ForkMain.java:300)
	at sbt.ForkMain$Run$2.call(ForkMain.java:290)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by patriknw** [comment]
Looks like it's via the `MiscMessageSerializer` -> `ActorRefResolver` -> `ActorSystemAdapter`, which calls `loadExtensions()`. Meaning that configured extensions are loaded too early.

**Comment by jrudolph** [comment]
Yep, exactly. In this case, it can be solved by making the `ActorRefResolver` lazy in `MiscMessageSerializer`.

I wonder though whether the (untyped) extension loading story is really solved in a principled way? It seems things might just happen to work because things are usually initialized in the right order. I guess it's ok this way. To unknot these kinds of dependency chains correctly seems to be a hard problem in general.

**Comment by patriknw** [comment]
yes, lazy can be needed in extensions that have cycles

here it feels somewhat more broken since `ActorSystemAdapter` that is an extension triggers the loading of the configured extensions, which is normally done as part of the AS startup phases. 
Perhaps `loadExtensions` could be triggered somewhat at the same time as the delayed guardian, @johanandren ?

**Comment by jrudolph** [comment]
There are so many things that could be considered broken if you just go digging into it:

 * untyped `ActorSystemImpl` initializes the actor ref provider first before initializing extensions, however, `RemoteActorRefProvider` already touches `SerializationExtension` during its initialization (which can lead to a chain of further extension loading
 * `ActorSystemAdapter` usually only runs after the untyped `ActorSystem` has been started but if its untyped extension is touched during untyped `ActorSystem` initialization it also might do so before
 * The `Cluster` extension tries to access internals from the `ClusterActorRefProvider`, which only works if the provider is done with initialization.
 * Dependency cycles between extensions might create deadlocks?

In the best case, initialization could be separated in phases, e.g. that no extensions are allowed to be accessed / initialized before at least the `ActorRefProvider` is initialized. Then there still might be cyclic dependencies between extensions but they can then be solved by sprinkling lazy vals where needed. Also, we should ensure that those cyclic dependencies are reported properly. It seems a cycle might actually dead lock because of the `await` in `ActorSystemImpl.findExtension`?

**Comment by johanandren** [comment]
Also note that I have an open PR that delays the user guardian startup, which is somewhat related: #24302

**Comment by jrudolph** [comment]
The last commit is basically a diagnostic help to prevent accidentally initializing things in the wrong order. It seems to break however in combination with the guardian.

**Review by patriknw — APPROVED** [review]
LGTM, I assume the `loadExtensions` is fixed elsewhere

**Comment by jrudolph** [comment]
No, I think, I'm actually also fixing the `loadExtension` as well. Failures are probably related.

**Comment by jrudolph** [comment]
I now fixed the `loadExtension` as well as suggested this morning by creating another untyped extension that is configured as a `library-extension` to load typed extensions as part of the untyped initialization process.

One notable change (also reflected in the last failing test) is that configured typed extensions are now always loaded also if just an untyped ActorSystem is created, as soon as the extension config is on the classpath. Consequently, also the typed ActorSystemAdapter is always created as soon as akka-actor-typed is on the classpath.

I guess that's fine because these operations are usually cheap enough.

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/ExtensionsSpec.scala [hunk 1]
```diff
@@ -151,17 +151,30 @@ class ExtensionsSpec extends TypedAkkaSpec {
...
       try {
-
         val before = InstanceCountingExtension.createCount.get()
         InstanceCountingExtension(untypedSystem.toTyped)
-        val ext = InstanceCountingExtension(untypedSystem.toTyped)
         val after = InstanceCountingExtension.createCount.get()
 
-        (after - before) should ===(1)
+        // should have been loaded even before it was accessed in the test because InstanceCountingExtension is listed
+        // as a typed library-extension in the config
```
Comment (johanandren) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/ExtensionsSpec.scala]: check the count before system was created as well, and see that system creation triggered one instantiation of it?

**Review by johanandren — COMMENTED** [review]
Goodie

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/ExtensionsSpec.scala (hunk 1 above)
Comment (jrudolph) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/ExtensionsSpec.scala]: Thanks good suggestion, changed.

**Pull request merged**